tests/rptest/tests/rack_aware_replica_placement_test.py

```diff
@@ -0,0 +1,116 @@
+        every available rack.
+        """
+        num_racks = len(set(rack_layout))
+        m = self.client().describe_topic(topic.name)
+        replicas = defaultdict(list)
+        for part in m.partitions:
+            for broker in part.replicas:
+                rack = rack_layout[broker - 1]  # broker ids are 1-based
+                replicas[part.id].append(rack)
+
+        for part_id, racks in replicas.items():
+            expected_racks = min(num_racks, num_replicas)
+            self.logger.info(
+                f"partition {part_id} is replicated to racks {racks}, expected {expected_racks} racks, total racks {num_racks}"
+            )
+            assert len(set(racks)) == min(num_racks, num_replicas)
+
+    @cluster(num_nodes=6)
+    @matrix(rack_layout_str=['AAAAAB', 'AABBCC', 'ABCDEF'],
+            num_partitions=[3, 30],
+            replication_factor=[3, 5])
+    def test_replica_placement(self, rack_layout_str, num_partitions,
+                               replication_factor):
+        """
+        Test replica placement. The test case creates a set of brokers with
+        initialized 'rack' parameter in the configuration. The 'rack' property
+        is defined by the 'rack_layout' parameter. Then it creates a topic with
+        pre-defined number of partitions and replication factor and validates
+        the replica placement.
+
+        The validation procedure checks that:
+        - either every replica lives on a different rack;
+        - or every rack is used to place one or more replicas of the partition 
+          (if replication factor is larger than number of racks)
+
+        ----
+        @param rack_layout_str is a 6-char string with all rack names combined into a signle string.
+                               Every character of the string corresponds to one broker. Every
+                               rack id has single character. 
+        @param num_partitions defines number of partitions that needs to be created.
+        @param replication_factor defines recplication factor of all partitions.
+        """
+
+        rack_layout = [str(i) for i in rack_layout_str]
+        assert len(rack_layout) == 6
+
+        for ix, node in enumerate(self.redpanda.nodes):
+            extra_rp_conf = {
+                # We're introducing two racks, small and large.
+                # The small rack has only one node and the
+                # large one has four nodes.
+                'rack': rack_layout[ix],
+                # This parameter enables rack awareness
+                'enable_rack_awareness': True,
+            }
+            self.redpanda.start_node(node, extra_rp_conf)
+
+        topic = TopicSpec(partition_count=num_partitions,
+                          replication_factor=replication_factor)
+
+        self._create_topic(topic)
+
+        # check that 'small' rack has 3 replicas
+        self._validate_placement(topic, rack_layout, replication_factor)
```

Comment:
would be useful if this test was wrapped in a loop and the verification occurred for 1 topic then 2 topics etc... just to make sure we don't have some sort of edge case that pops up outside of a single topic test?

Reply:
also fixed